### ISSUE TITLE
[BUG] Path to FFMPEG No Longer Updatable in 10.8.13

### ISSUE THREAD

**Issue by a-priestley** [linuxserver/docker-mods#899]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Name of mod

jellyfin-rffmpeg

### Name of base container

jellyfin

### Current Behavior

As of 10.8.13, this mod is no longer usable due to the inability for users to manually set the path to the ffmpeg binary.

[See here](https://jellyfin.org/posts/jellyfin-security-and-you/)

### Expected Behavior

The way by which the mod operates may have to change. Setting the value in encoding.xml followed by a restart does not seem to work either. Possibly due to:

> Note too that the --ffmpeg flag may or may not change this for you, depending on what version you originally installed; while you should change it to match, you may still need to update the XML file as well.

Note that in the `run` script of `docker-jellyfin`, the flag mentioned above is hard coded, possibly overwriting the value entered into the .xml:

```sh
#!/usr/bin/with-contenv bash
# shellcheck shell=bash

export \
    JELLYFIN_DATA_DIR="/config/data" \
    JELLYFIN_CONFIG_DIR="/config" \
    JELLYFIN_LOG_DIR="/config/log" \
    JELLYFIN_CACHE_DIR="/config/cache" \
    JELLYFIN_WEB_DIR="/usr/share/jellyfin/web"

exec \
    s6-notifyoncheck -d -n 300 -w 1000 \
        s6-setuidgid abc /usr/bin/jellyfin \
        --ffmpeg=/usr/lib/jellyfin-ffmpeg/ffmpeg
```

### Steps To Reproduce

1. Running with the following docker-compose.yml service: 
```yml
services:
  jellyfin:
    image: lscr.io/linuxserver/jellyfin:arm64v8-10.8.13
    network_mode: host
    container_name: jellyfin
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Etc/UTC
      - JELLYFIN_PublishedServerUrl=192.168.0.120
      - DOCKER_MODS=linuxserver/mods:jellyfin-rffmpeg
      - RFFMPEG_USER=jellyfin
      - RFFMPEG_HOST=100.64.0.1
      - RFFMPEG_HOST_MAC="50:eb:f6:7d:e9:f7"
      - RFFMPEG_WOL=native
      - FFMPEGDIR=/usr/local/bin/wol_rffmpeg/ffmpeg
      - WOL_NATIVE_HOST=192.168.2.101
      - WOL_NATIVE_PORT=9999
      - WOL_WAIT=10
    volumes:
      - ${PATH_TO_CONFIG}/jellyfin:/config
      - /shares/docker-data/jellyfin/data/subtitles:/config/data/data/subtitles
      - /shares/docker-data/jellyfin/transcodes:/config/data/transcodes
      - /shares/media:/data
    restart:
      unless-stopped
```
2. Login as administrator
3. Navigate to dashboard > playback > transcoding
4. The `FFmpeg path` is no longer an editable field
5. Note that the following fields in `/config/encoding.xml` are already set:
```xml
...
  <EncoderAppPath>/usr/local/bin/wol_rffmpeg/ffmpeg</EncoderAppPath>
  <EncoderAppPathDisplay>/usr/lib/jellyfin-ffmpeg/ffmpeg</EncoderAppPathDisplay>
...
```
6. Restart the service. Observe that the specified path is still not used, and jellyfin uses the default binary to transcode media.

### Environment

```markdown
- OS: Alpine Raspberry PI
- How docker service was installed: `docker-compose` installed via `apk`
```

### CPU architecture

arm64

### Docker creation

```bash
See above.
```

### Container logs

```bash
...
[custom-init] No custom files found, skipping...
[21:49:44] [INF] [1] Main: Jellyfin version: 10.9.1
[21:49:44] [INF] [1] Main: Environment Variables: ["[JELLYFIN_WEB_DIR, /usr/share/jellyfin/web]", "[JELLYFIN_DATA_DIR, /config/data]", "[JELLYFIN_LOG_DIR, /config/log]", "[JELLYFIN_PublishedServerUrl, 192.168.0.120]", "[JELLYFIN_CACHE_DIR, /config/cache]", "[JELLYFIN_CONFIG_DIR, /config]"]
[21:49:44] [INF] [1] Main: Arguments: ["/usr/lib/jellyfin/bin/jellyfin.dll", "--ffmpeg=/usr/lib/jellyfin-ffmpeg/ffmpeg"]
[21:49:44] [INF] [1] Main: Operating system: Ubuntu 22.04.4 LTS
[21:49:44] [INF] [1] Main: Architecture: Arm64
[21:49:44] [INF] [1] Main: 64-Bit Process: True
[21:49:44] [INF] [1] Main: User Interactive: True
[21:49:44] [INF] [1] Main: Processor count: 4
[21:49:44] [INF] [1] Main: Program data path: /config/data
[21:49:44] [INF] [1] Main: Log directory path: /config/log
[21:49:44] [INF] [1] Main: Config directory path: /config
[21:49:44] [INF] [1] Main: Cache path: /config/cache
[21:49:44] [INF] [1] Main: Web resources path: /usr/share/jellyfin/web
[21:49:44] [INF] [1] Main: Application directory: /usr/lib/jellyfin/bin/
...
[21:50:06] [INF] [1] [21:50:05] [INF] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Available filters: ["overlay_opencl", "overlay_rkrga", "scale_opencl", "scale_rkrga", "tonemap_opencl", "vpp_rkrga", "zscale", "alphasrc"]
[21:50:05] [WRN] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Filter: scale_cuda with option Output format (default "same") is not available
[21:50:05] [WRN] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Filter: tonemap_cuda with option GPU accelerated HDR to SDR tonemapping is not available
[21:50:06] [WRN] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Filter: overlay_vaapi with option Action to take when encountering EOF from secondary input is not available
[21:50:06] [WRN] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Filter: overlay_vulkan with option Action to take when encountering EOF from secondary input is not available
[21:50:06] [INF] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Available hwaccel types: ["drm", "opencl", "rkmpp"]
[21:50:06] [INF] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: FFmpeg: /usr/lib/jellyfin-ffmpeg/ffmpeg
...
```
```

**Comment by junkman690** [comment]
Jellyfin not picking up the path correctly from the xml seems like a Jellyfin bug which has popped up across RPM and Free-BSD builds, however its been completely removed in 10.9 https://jellyfin.org/posts/jellyfin-release-10.9.0#transcoding--ffmpeg. Agree the mod will need to change from relying on end user changes to setting the --ffmpeg flag to an environment variable. I have changed my setup since creating this mod and no longer actively use it. I will see if I can find some time to update but happy for someone else to take that on in the meantime.

**Comment by a-priestley** [comment]
I can make the needed changes. Are there any good examples for altering the run script? My initial thought is to edit the `--ffmpeg` flag based on environment variables. Default to `/usr/local/bin/rffmpeg`. If wol is needed you can assign the variable to change.

**Comment by a-priestley** [comment]
I've added the svc directory with the following script with `chmod +x`:
```sh
#!/usr/bin/with-contenv bash
# shellcheck shell=bash

export \
    JELLYFIN_DATA_DIR="/config/data" \
    JELLYFIN_CONFIG_DIR="/config" \
    JELLYFIN_LOG_DIR="/config/log" \
    JELLYFIN_CACHE_DIR="/config/cache" \
    JELLYFIN_WEB_DIR="/usr/share/jellyfin/web"

exec \
    s6-notifyoncheck -d -n 300 -w 1000 \
        s6-setuidgid abc /usr/bin/jellyfin \
        --ffmpeg=/usr/local/bin/rffmpeg/wol_rffmpeg/ffmpeg
```

I've hard coded the flag here to test on my environment with the type file containing "longrun", but it hasn't worked. I still see the default path in my jellyfin logs.

Is the svc script supposed to overwrite the main jellyfin run script, or does anything additional need to be done?

**Comment by thespad** [comment]
If you want a mod to overwrite the built-in init files you need to place them in exactly the same locations so that when the mod is unpacked it replaced them.

**Comment by a-priestley** [comment]
> If you want a mod to overwrite the built-in init files you need to place them in exactly the same locations so that when the mod is unpacked it replaced them.

Thank you! Working now.

I suppose I can import `FFMPEGDIR` into this script so that it can be manually set. At the same time, it might be better to remove the flag entirely from the main linuxserver/jellyfin container? I think this variable is supposed to set the path in the official release, and the flag is overwritting it.

**Comment by thespad** [comment]
According to the 10.9 release notes

> The --ffmpeg command-line flag is now the primary method of setting FFmpeg paths, and configuration of the FFmpeg paths via encoding.xml is now deprecated. All official packaging will revert to defaults unless explicitly set before upgrade.

**Comment by a-priestley** [comment]
Do docker env vars get imported into the run script by default? Can't seem to get mine to take:
```sh
#!/usr/bin/with-contenv bash
# shellcheck shell=bash

if [ -z "$RFFMPEG_PATH" ]
then
    RFFMPEG_PATH=/usr/local/bin/ffmpeg
fi

export \
    JELLYFIN_DATA_DIR="/config/data" \
    JELLYFIN_CONFIG_DIR="/config" \
    JELLYFIN_LOG_DIR="/config/log" \
    JELLYFIN_CACHE_DIR="/config/cache" \
    JELLYFIN_WEB_DIR="/usr/share/jellyfin/web"

exec \
    s6-notifyoncheck -d -n 300 -w 1000 \
        s6-setuidgid abc /usr/bin/jellyfin \
        --ffmpeg="$RFFMPEG_PATH"
```

By the way this is my first time editing one of these. If there's a better way, let me know.

**Comment by a-priestley** [comment]
Disregard the bit about the variable. I must have forgotten to write after editing my yaml.

Second question still valid! :)

**Comment by aptalca** [comment]
Maybe we should add that var to the jellyfin image as an unlisted one and this mod can just set it. No need to override the whole service file

**Comment by thespad** [comment]
If we're going to do that we probably also want to add an `-f` check for the supplied path to make sure they're pointing to a file that actually exists.

**Comment by a-priestley** [comment]
> Maybe we should add that var to the jellyfin image as an unlisted one and this mod can just set it. No need to override the whole service file

I agree with this. The readme for this mod will need to be updated to reflect it as well.



> If we're going to do that we probably also want to add an `-f` check for the supplied path to make sure they're pointing to a file that actually exists.
> 
> And it'll need to be documented by the mod because there are two different RFFMPEG binaries to pick from.

This also makes sense. Like this:
```sh
#!/usr/bin/with-contenv bash
# shellcheck shell=bash

if [[ -z "$FFMPEG_PATH" || -f "$FFMPEG_PATH" ]]
then
    FFMPEG_PATH=/usr/lib/jellyfin-ffmpeg/ffmpeg
fi

export \
    JELLYFIN_DATA_DIR="/config/data" \
    JELLYFIN_CONFIG_DIR="/config" \
    JELLYFIN_LOG_DIR="/config/log" \
    JELLYFIN_CACHE_DIR="/config/cache" \
    JELLYFIN_WEB_DIR="/usr/share/jellyfin/web"

exec \
    s6-notifyoncheck -d -n 300 -w 1000 \
        s6-setuidgid abc /usr/bin/jellyfin \
        --ffmpeg="$FFMPEG_PATH"
```

This would be for the main jellyfin image, and the mod readme would simply need to be updated to reflect this, with the two options available.

**Comment by thespad** [comment]
Happy side effect from https://github.com/linuxserver/docker-jellyfin/pull/248 and https://github.com/linuxserver/docker-jellyfin/pull/249 is that I realised we never added the proper health endpoint check to the nightly branch.

**Comment by a-priestley** [comment]
#900

**Comment by junkman690** [comment]
I was thinking on this last night, only the wol path needs to be configured and adding an additional check that if the wol env var isn't set to native or API it just kicks off rffmpeg. Readme would still need to be updated to reflect as well.

**Comment by a-priestley** [comment]
I think users would still want the option of providing a custom path even if they were using the mod for flexibility reasons.
Maybe we could check the var and set it to normal rffmpeg if not set?

**Comment by junkman690** [comment]
Agree ability to set a custom is good, but looks like that's being added to the main container. The wol version is only a minor shell script, and might make it easier to use only having it specified rather than giving two different possible paths in the readme.

My original thought was to just sed the run script to change the -ffmpeg flag but giving it as an option in the main container give more flexibility.

**Comment by junkman690** [comment]
Actually, re-reading the wol script it already accommodates just running rffmpeg if API or native isn't set.

**Comment by thespad** [comment]
Both Jellyfin branch PRs have merged now, providing the option of passing a `FFMPEG_PATH` to the service. In your case probably via the mod.

**Comment by thespad** [comment]
Just be wary of anyone still running old versions who will still be pulling the mod and try not to break things for them.

**Comment by thespad** [comment]
This should now be resolved by #901
